transactions/transactor.go

```diff
@@ -165,6 +145,82 @@ func (t *Transactor) SendTransactionWithSignature(args SendTxArgs, sig []byte) (
 	return signedTx.Hash(), nil
 }
 
+func (t *Transactor) HashTransaction(args SendTxArgs) (validatedArgs SendTxArgs, hash gethcommon.Hash, err error) {
```

Comment:
i think it introduces several concurrency issues
- two concurrent transactions generated with this method will use same nonce
- local nonce won't get incremented if transaction generated with this method is submitted on chain

Reply:
@dshulyak you are right, but in this specific case status-react will:

1 - call `HashTransaction`
2 - let the keycard sign the hash
3 - call `SendTransactionWithSignature`

Step 3 will increment the nonce. 
In theory from the app UI you won't be able to do anything between step 1 and 2.
And also, `SendTransactionWithSignature` checks if the transaction that we are sending has the expected nonce, otherwise it fails, so we shouldn't have problem.

Reply:
Regarding concurrent `HashTransaction` calls with the same address, won't `t.addrLock.LockAddr(args.From)` prevent from using the same nonce?

Reply:
Generally, I think that the flow with the keycard is fine. It's on the caller to make sure it does not execute concurrent transactions using `HashTransaction`.

Reply:
>  ok, I see the problem. They don't need to be concurrent actually to use the same nonce.

why?

as i understand nonce is incremented in another operation called `SendTransactionWithSignature`, so the concurrency issue that i mentioned is a following one:
- HASH tx1
- HASH tx2
- SendTransactionWithSignature tx1
- SendTransactionWithSignature tx2

both tx1 and tx2 will use same nonce

i don't mind if whole flow is controlled by a client, btw there is no way to wait for this `2 - let the keycard sign the hash` in status-go?

Reply:
yes you are right, it would be a problem meaning that tx1 and tx2 would have the same nonce but 
 SendTransactionWithSignature would fail with the second one. 
But as we said the process is managed by the client so from the UI you won't be able to do anything in between

Reply:
yes, i understood that from the first comment :). and i have no issues with that. i was mostly trying to clarify why @adambabik said that tx's don't need to be concurrent